Combine locus keyed

cseed opened the pull request:
First, I changed import_vcfs to return a MatrixTable only keyed by locus, and removed the MatrixKeyRowsBy in combine_gvcfs.  To goal here is to avoid re-buidling an re-broadcasting the partitioner once for each gVCF.  We'll need to re-key at the very end.  I'm not so familiar with the end of the joint calling pipeline.  @chrisvittal can you take care of that?

Second, I don't repartition in TableMultiWayZipJoin if the partitioners all match (which they should in in the joint calling pipeline).  For that to work right, I need allowedOverlap == 0 (or to verify the partitions are in fact disjoint).  Turns out allowedOverlap wasn't being propagated in various places.  I fixed that.

@patrick-schultz can you look at the RVDPartitioner changes?  They just look like oversights to me, but maybe there was a reason why, for example, copy and coarsen wasn't preserving allowedOverlap?

Finally, now the joint calling pipeline/test_combiner_works segfaults, ugh:

```
$ hail -m unittest test.hail.methods.test_impex.VCFTests.test_combiner_works
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000010e5fa090, pid=64905, tid=33795
#
# JRE version: Java(TM) SE Runtime Environment (8.0_45-b14) (build 1.8.0_45-b14)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.45-b02 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# J 8877 C1 is.hail.expr.types.physical.PLocus$$anon$1.compare(Lis/hail/annotations/Region;JLis/hail/annotations/Region;J)I (117 bytes) @ 0x000000010e5fa090 [0x000000010e5f9de0+0x2b0]
#
```

The rest of the tests pass (the other Python failures are cascaded failures from test_combiner_works, I double-checked in the hopes of finding an easier example to debug.)  It is pretty clearly related to the no repartition optimization.  If I disable it, test_combiner_works passes.

I haven't tracked this down, but I do have one question @chrisvittal: who's responsible for freeing the inputs (that is, clearing the input regions) to multi-way zip join?  I don't see where that happens.

cseed commented:
> who's responsible for freeing the inputs

I think that was it.  When I put `.boundary` on the inputs (which frees the previous element when getting the next one) it works: https://github.com/hail-is/hail/pull/5424/commits/bdc7fb0a3dd4afc6b69da56542747a7b33d23e81.  We were probably just running out of memory.

cseed commented:
update: I made checking the keys in the RVD constructor optional, as it forced the broadcast of the partitioner every time we construct an RVD, and enabled it (via a flag HailContext.checkRVDKeys) only for the Scala tests.

@tpoterba Thoughts on this?  Should keep it enabled in production and just add a flag to disable it for the joint caller?  If not, should we enable it for the Python tests?  If so, we'll probably want some additional tests with it disabled so we at least test somewhat we release.

cseed commented:
@chrisvittal I verified the partitioner is no longer broadcast for the combiner pipeline.  This should be ready for review.

chrisvittal commented on hail/python/hail/experimental/vcf_combiner.py:
```diff
@@ -129,10 +128,7 @@ def min_rep(locus, ref, alt):
     combined = combine(ts)
     combined = combined.annotate(alleles=fix_alleles(combined.alleles))
     return hl.MatrixTable(
```
If we don't need the `MatrixKeyRowsBy` then we don't need to construct a new `MatrixTable`, `_unlocalize_entries` will take care of it.

cseed replied:
Ah, yes, good catch, fixed.

patrick-schultz commented on hail/src/main/scala/is/hail/rvd/RVDPartitioner.scala:
```diff
@@ -13,13 +13,13 @@ class RVDPartitioner(
   // rangeBounds is interval containing all keys within a partition
   val rangeBounds: Array[Interval],
-  allowedOverlap: Int
+  val allowedOverlap: Int
```
`allowedOverlap` was intentionally not a field of the partitioner class. It represents a property that the `rangeBounds` may or may not have, and by passing it as an argument to the constructor, the caller can ask the constructor to enforce that the passed `rangeBounds` satisfy that property.

But I never saw any use cases for saving the `allowedOverlap` argument. You can always check if a partitioner satisfies a certain `allowedOverlap` with the `satisfiesAllowedOverlap` method. I don't understand what the `allowedOverlap` check is doing in `TableMultiWayZipJoin`, so maybe you did find a use case. Could you explain what you're trying to do?

patrick-schultz replied:
>  For that to work right, I need allowedOverlap == 0 (or to verify the partitions are in fact disjoint).

Ah, based on this, I think `rvd.partitioner.satisfiedAllowedOverlap(0)` does what you want.

patrick-schultz replied:
> Oh, allowedOverlap isn't even used, except in the assertion. Should we even keep it in?

Right, it's just asking for a stronger precondition to be checked on the range bounds during construction. Personally, I think it's worth keeping for that purpose.

cseed commented on hail/src/main/scala/is/hail/expr/ir/TableIR.scala:
```diff
@@ -688,14 +688,23 @@ case class TableMultiWayZipJoin(children: IndexedSeq[TableIR], fieldName: String
+    val repartitionedRVDs =
+      if (childRVDs.forall(rvd =>
+        (rvd.partitioner.rangeBounds sameElements childRVDs(0).partitioner.rangeBounds) &&
+          rvd.partitioner.allowedOverlap == 0))
```
I'm going to zip together a bunch of different inputs.  So, if the partitions overlap, it may be some keys in this partition are in another, overlapping partition in another input, no?  So I want to check there is no overlap.

cseed commented on hail/src/main/scala/is/hail/expr/ir/TableIR.scala:
```diff
@@ -688,14 +688,23 @@ case class TableMultiWayZipJoin(children: IndexedSeq[TableIR], fieldName: String
+          rvd.partitioner.allowedOverlap == 0))
+        childRVDs
+      else {
+        info("TableMultiWayZipJoin: repartitioning children")
```
> `repartition` is almost a no-op

`RepartitionedOrderedRDD2` does a partitioner broadcast which is not viable in my setting.  That's what I'm trying to eliminate.  Moreover, I'm trying to merge 120K files together, so "almost a no-op" is multiplied by 120K, so I'm quite wary of almost in this setting.

I'm also slightly concerned about comparing all the partitioners: that's O(NM) where N is the number of inputs (= 120K) and M is the partitioner size (~20K).  In my case, they partitions are all the same (eq the same), so if I keep the code like it is but make sure RVDPartitioner.equals short circuits on eq, then that term goes away, but I can't do that via generate/repartition.  Hmm.

cseed replied:
OK, modified it to use satisfiesAllowedOverlap and compare using RVDPartitioner.equals which short circuits on ref equality (and added RVDPartitioner.hashCode.  You have to define hashCode when you define equals.)

I'm not quite sure what to do about the O(NM) and putting the optimization in repartition.

cseed replied:
Also, RVDPartitioner.equals was using == on arrays, which is ref equality.  I assume that wasn't intended?

patrick-schultz replied:
> Also, RVDPartitioner.equals was using == on arrays, which is ref equality. I assume that wasn't intended?

Nope, that should have been `sameElements`.

cseed commented:
OK, I left in the optimization in multi-way join but also added it to RVD.repartition.  This is the best I can see how to do.

I realize I wanted `satisfiesAllowedOverlap(key.length - 1)`, not `satisfiesAllowedOverlap(0)`.  It's just the case I was working with had one key.  Fixed.  That might clarify some confusion.

Also added a strictify to generate.  I think the existing code was wrong without that.  @patrick-schultz can you take another look?

patrick-schultz commented:
> I realize I wanted `satisfiesAllowedOverlap(key.length - 1)`, not `satisfiesAllowedOverlap(0)`.

Because `allowedOverlap` is tricky to think about, I also gave constructor overloads that take `partitionKey` like in the old style. So this case is equivalent to setting `partitionKey == key` in the constructor `def this(partitionKey: Array[String], kType: TStruct, rangeBounds: IndexedSeq[Interval])`, if you think that's clearer.

patrick-schultz commented on hail/src/main/scala/is/hail/rvd/RVDPartitioner.scala:
```diff
@@ -184,9 +194,10 @@ class RVDPartitioner(
+    allowedOverlap: Int = allowedOverlap
   ): RVDPartitioner =
-    new RVDPartitioner(kType, rangeBounds)
+    new RVDPartitioner(kType, rangeBounds, allowedOverlap)
```
This `allowedOverlap` and the one in `coarsen` should be removed, so that `allowedOverlap` doesn't get added as a field.

cseed replied:
> so that `allowedOverlap` doesn't get added as a field

I removed `allowedOverlap` as a field (that is, it is no longer a `val`).  `copy` is a convenience constructor.  Shouldn't have the same arguments as the constructor and preserve the arguments that aren't overridden?

Isn't `coarsen` preserving the overlap (up to truncation)?  Why shouldn't it correctly propagate that information forward?

Maybe I don't know what you mean by "allowedOverlap" is a field here.

patrick-schultz replied:
I just mean that if `allowedOverlap` is only used in the (implicitly defined) constructor, then Scala doesn't need to store it in a field in the generated Java class.

`allowedOverlap` doesn't have any information content that needs to be propagated. Either the `rangeBounds` satisfy the property or they don't. Including `allowedOverlap` in `copy` and `coarsen` is just requesting an extra check during construction of the new partitioner that is impossible to fail. My intention for leaving `allowedOverlap` as a constructor argument was purely as a more convenient alternative to
```
assert(RVDPartitioner.isValid(kType, rangeBounds, allowedOverlap))
val partitioner = new RVDPartitioner(kType, rangeBounds)
```

cseed replied:
Ah, you mean you don't want to reference it outside the constructor in which case it will actually get stored?

I think I'm starting to see your logic.  `allowedOverlap` is so clients outside RVDPartitioner can specify the expected allowed overlap and get it checked?  I guess I have two questions: (1) Does that ever happen?  Who knows their allowedOverlap statically in an interesting way?  And (2) Why are you including internal uses of the RVDPartitioner constructor?  Shouldn't they get the benefit of that checking, too?  (Like coarsen?)

cseed replied:
> Including `allowedOverlap` in `copy` and `coarsen` is just requesting an extra check during construction of the new partitioner that is impossible to fail.

Unless there is a bug in coarsen or the caller of copy.  Isn't that exactly the kind of thing `allowedOverlap` is there to check for?

patrick-schultz replied:
I don't think it would take much to convince me to remove it completely. I left it there as a bridge to the old `partitionKeys` world.

1) As an example, `LoadVCF` calls `new RVDPartitioner(Array("locus"), fullType.rowKeyStruct, rangeBounds.asInstanceOf[IndexedSeq[Interval]])`, which dispatches to the primary constructor using the `allowedOverlap` argument to enforce that the partitioner being built is consistent with the old "partitionKey = locus" semantics.
2) Internal uses are easier to prove corectness. For example, if `rangeBounds` satisfies `allowedOverlap = n`, then of course it still will in the new partitioner made by `copy`. Whatever `allowedOverlap` should be passed to the constructor in `coarsen` should likewise be provably guaranteed to pass. Maybe storing `allowedOverlap` in the java object in order to get that extra runtime sanity check is worth it. I was less concerned about whether `allowedOverlap` gets stored than I was making clear that it is not semantically an attribute of the partitioner.

cseed replied:
I've never understood "impossible to fail" as a complaint against assertions.  All asserts should be impossible to fail by definition.  It's only the bugs that makes them not so, and that's what you're looking for.

Anyway, I don't really care to fight about this.  With `satisfiesAllowedOverlap`, my case is handled.  I'll back out the changes if you prefer.

patrick-schultz replied:
> Unless there is a bug in coarsen or the caller of copy. Isn't that exactly the kind of thing `allowedOverlap` is there to check for?

Yeah, you've convinced me. Ignore this request.

patrick-schultz replied:
Sorry, I didn't want a fight either. Just trying to explain my original reasoning. You're completely right about assertions.

cseed replied:
OK, great.

 > I was less concerned about whether `allowedOverlap` gets stored than I was making clear that it is not semantically an attribute of the partitioner.

So it feels like a semantic attribute to me, it is just a (very cheap!)  bound on the property you can compute exactly from rangeBounds in O(N).  But I see now where you were coming from.  Thanks for discussing!

cseed commented on hail/src/main/scala/is/hail/expr/ir/TableIR.scala:
```diff
@@ -696,6 +696,7 @@ case class TableMultiWayZipJoin(children: IndexedSeq[TableIR], fieldName: String
         info("TableMultiWayZipJoin: repartitioning children")
         val childRanges = childRVDs.flatMap(_.partitioner.rangeBounds)
         val newPartitioner = RVDPartitioner.generate(childRVDs.head.typ.kType.virtualType, childRanges)
+          .strictify
```
Yes, good, removed.

patrick-schultz commented:
I'm ready to approve. @chrisvittal?

chrisvittal commented:
I feel that I've learned more about RVDs and partitioners. The front end of this change looks correct. '+1' from me.

cseed commented:
> I feel that I've learned more about RVDs and partitioners.

Same here!  Good discussion!

catoverdrive commented:
@cseed I'd like to turn on the key checking again for the time being because we were relying on it for some split_multi stuff that's going to take some amount of new infrastructure to fix properly, and in the meantime it's causing some pretty bad/nonsense errors downstream because things are can be out of order (see #6223). Is that going cause any problems?

catoverdrive commented:
Tim pointed out that we probably only need to check when we're passing `isSorted` = True in TableKeyBy, so we can just add a check there
